### PR TITLE
add resource consumption testing result

### PR DESCRIPTION
**Pull request by PerforMance308**
### Collected CPU usage metrics in two scenarios:
- Users' pod has CPU limit
- Users' pod doesn't have CPU limit

### Collected Memory consumption

### CODE REVIEW

File: content/en/docs/performance/resource_consumption.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+menu:
+  docs:
+    parent: performance
+    weight: 2
+title: Kmesh resource consumption
+toc: true
+type: docs
+---
+# 1. Environment setup
+![resource_env](images/resource_test_env.png)
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:13]: Please check all pictures, not rendered

File: content/en/docs/performance/resource_consumption.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+Set the CPU limit for App A to 1 (1 CPU), and collect corresponding Kmesh eBPF CPU consumption.
+
+**Note: With the system having 8 cores and a CPU limit of 1, it means the POD can consume up to 12.5% of the CPU.**
+
+Testing result：
+![resource_result1](images/resource_test_result1.png)
+
+**Results and conclusions: When APP A uses up 1 CPU, the user space occupies 4%, the kernel space occupies 7.4%, totaling 11.4% of CPU consumption, which is less than the POD CPU limit of 12.5%. There are two possible reasons.**
+
+- APP A and Kmesh eBPF share the POD CPU limit, with Kmesh eBPF CPU being restricted by the POD CPU limit.
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:36]: why ebpf share pod cpu usage with pod?  ebpf is run in kernel, whild pod cpu usage includes all the process cpu usage within the pod

File: content/en/docs/performance/resource_consumption.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+for (i;0;i<10000;i++) {}
+```
+
+Testing result：
+![resource_result3](images/resource_test_result3.png)
+
+When APP A fully utilizes 1 CPU, the user space consumes 3.9% CPU, while the kernel space consumes 8.3%, totaling 12.2% CPU usage, which is still below the POD CPU limit (12.5%). Through multiple rounds of testing, the total CPU consumption of APP A and eBPF is always less than the POD CPU limit.
+
+
+**Conclusion: Kmesh eBPF and APP share the POD CPU limit, with Kmesh eBPF CPU being limited by the POD CPU limit.**
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:64]: Not correct

File: content/en/docs/performance/resource_consumption.md (hunk 2 above)
Comment (PerforMance308) [reply content/en/docs/performance/resource_consumption.md:36]: we have presented and discussed this testing result in the weekly meeting with @nlgwcy , and everybody are aware of this.

File: content/en/docs/performance/resource_consumption.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+type: docs
+---
+# 1. Environment setup
+![resource_env](images/resource_test_env.png)
+
+- K8S version: v1.27
+- Kmesh version: 0.4
+- Kernel version: 5.10
+- Node flavor: 8U16G
+- testing tool: fortio
```
Comment (YaoZengzeng) [inline content/en/docs/performance/resource_consumption.md:19]:
```suggestion
- Testing tool: fortio
```

File: content/en/docs/performance/resource_consumption.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+- metric collecting tools: top, bpftop
```
Comment (YaoZengzeng) [inline content/en/docs/performance/resource_consumption.md:20]:
```suggestion
- Metric collecting tools: top, bpftop
```

File: content/en/docs/performance/resource_consumption.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+When App A consumes 1 CPU, the user space occupies 4.4%, the kernel space occupies 7.7%, totaling 12.1% CPU usage. This is less than the POD CPU limit of 12.5%.
+
+**Results and Conclusion: When App A consumes 1 CPU, the user space occupies 4.4%, the kernel space occupies 7.7%, totaling 12.1% CPU usage, which is less than the POD CPU limit of 12.5%. To further validate this conclusion, Experiment 2.3 will be conducted.**
+
+## 2.3 Based on test 2.2, modify the eBPF code to decrease its performance, causing it to consume more CPU. Observe if it can exceed the POD CPU limit.
+
+Add for loop in Kmesh eBPF code：
+```c
+for (i;0;i<10000;i++) {}
```
Comment (YaoZengzeng) [inline content/en/docs/performance/resource_consumption.md:55]: Won't the compiler optimize these code away?

File: content/en/docs/performance/resource_consumption.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+## 4.1 Create 1, 100, and 1000 services respectively in the cluster, and record the eBPF memory consumption.
+Testing result:
+|service number|eBPF Memory usage|
+|--|--|
+|1|23m|
+|100|23m|
+|1000|23m|
+
+**Test results: The eBPF memory consumption of Kmesh is 23MB and remains constant, regardless of the number of services.**
```
Comment (YaoZengzeng) [inline content/en/docs/performance/resource_consumption.md:119]: Are these services associated with real pods? Won't the ebpf map grow as the number of servcies grow?

File: content/en/docs/performance/resource_consumption.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+![resource_env](images/resource_test_env.png)
+
+- K8S version: v1.27
+- Kmesh version: 0.4
+- Kernel version: 5.10
+- Node flavor: 8U16G
+- testing tool: fortio
+- metric collecting tools: top, bpftop
+
+
```
Comment (YaoZengzeng) [inline content/en/docs/performance/resource_consumption.md:22]: I think it's necessary to add an overall introduction to the purpose of this article

File: content/en/docs/performance/resource_consumption.md (hunk 7 above)
Comment (PerforMance308) [reply content/en/docs/performance/resource_consumption.md:119]: Those are mock services, but I can see the data in the bpf Map. no change in the size of eBPF map

File: content/en/docs/performance/resource_consumption.md (hunk 6 above)
Comment (PerforMance308) [reply content/en/docs/performance/resource_consumption.md:55]: According to the test there is no optimization

**Comment by PerforMance308** [comment]
> This tests need some improvements
> 
> 1. First lack of how the test is run, assume this is ads mode
> 2. The cpu usage monitor is very corse, should use pod monitor
> 3. About the memory test, I cannot get how you get the usage. Should supplement the procedure,
> 
> For the cpu test, should distinguish different conccurrent connections, qps

the purpose to conduct this test is to get the Kmesh maximum CPU and momery usage, and see if there are any limitations of that. So the connections, qps metrics are not considered in this test

File: content/en/docs/performance/resource_consumption.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+title: Kmesh resource consumption
+toc: true
+type: docs
+---
+# 1. Background Information
+eBPF programs consume a certain amount of CPU and memory resources during their execution, which may affect the performance of other applications. To gain a deeper understanding of Kmesh’s resource consumption under different loads, we conducted several CPU and memory stress tests. These tests aim to evaluate the limits of CPU and memory consumption of Kmesh in real-world usage scenarios.
+
+**This documentation is based on Kmesh 0.4 ads mode**
+# 2. Environment setup
+![resource_env](images/resource_test_env.png)
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:17]: Still not able to show https://deploy-preview-67--kmesh-net.netlify.app/en/docs/performance/resource_consumption/#31-deploy-1-app-a-and-set-app-a-pods-cpu-limitthen-generage-load-and-collect-kmesh-ebpf-cpu-usage

Please refer to other docs

File: content/en/docs/performance/resource_consumption.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+```sh
+$ ./bpftop
+```
+
+Testing result：
+![resource_result1](images/resource_test_result1.png)
+
+the 12.99% in the figure above means that 12.99% of one CPU core was consumed.
+
+**Results and Conclusion: When APP A fully utilizes one CPU core, the eBPF program consumes 1.73% of the CPU, which is less than the POD CPU limit of 12.5%. There are two possible reasons.**
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:53]: how does 1.73b come out

File: content/en/docs/performance/resource_consumption.md [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+the 13.42% in the figure above means that 13.42% of one CPU core was consumed.
+
+
+**Results and Conclusion: When APP A fully utilizes one CPU, the eBPF program consumes 1.81% of the CPU, which is less than the POD CPU limit of 12.5%. To further verify this conclusion, experiment 3.3 will be conducted.**
+
+## 3.3 Based on test 2.2, modify the eBPF code to decrease its performance, causing it to consume more CPU. Observe if it can exceed the POD CPU limit.
+
+Add for loop in Kmesh eBPF code：
+```c
+for (i=0;i<65535;i++) {
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:75]: I am not sure how does this correslate to kmesh

File: content/en/docs/performance/resource_consumption.md (hunk 9 above)
Comment (PerforMance308) [reply content/en/docs/performance/resource_consumption.md]: I changed the image path, but the image still doesn't show up

**Review by hzxuzhonghu — COMMENTED** [review]
The graph does not show up can you reference other article

This is the preview https://deploy-preview-67--kmesh-net.netlify.app/en/docs/performance/resource_consumption/

File: content/en/docs/performance/resource_consumption.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+menu:
+  docs:
+    parent: performance
+    weight: 2
+title: Kmesh Resource Consumption
+toc: true
+type: docs
+---
+# 1. Background Information
+eBPF programs consume a certain amount of CPU and memory resources during their execution, which may affect the performance of other applications. To gain a deeper understanding of Kmesh’s resource consumption under different loads, we conducted several CPU and memory stress tests. These tests aim to evaluate the limits of CPU and memory consumption of Kmesh in real-world usage scenarios.
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:13]: This is a small amount, i would suggest removing `which may effect the performance of other applications`

File: content/en/docs/performance/resource_consumption.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+**This documentation is based on Kmesh 0.4 ads mode**
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:15]: now we only call kernel-native mode

File: content/en/docs/performance/resource_consumption.md [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+- Kmesh version: 0.4 ads mode
+- Kernel version: 5.10
+- Node flavor: 8U16G
+- Testing tool: fortio
+- Metric collecting tools: [bpftop](https://github.com/Netflix/bpftop), [inspektor-gadget](https://github.com/inspektor-gadget/inspektor-gadget)
+
+
+# 3. Test case 1 - POD with CPU limit scenario，collect Kmesh eBPF CPU usage
+## 3.1 Deploy 1 App A and set App A POD's CPU limit，then generage load and collect Kmesh eBPF CPU usage
+![resource_test1](/docs/performance/resource_test1.png)
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:29]:
```suggestion
![resource_test1](/docs/performance/images/resource_test1.png)
```

File: content/en/docs/performance/resource_consumption.md [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+
+Testing result：
+![resource_result1](/docs/performance/resource_test_result1.png)
+
+the 12.99% in the figure above means that 12.99% of one CPU core was consumed.
+
+**Results and Conclusion: When APP A fully utilizes one CPU core, the eBPF program consumes 1.73%(totally 13.9% usage of one CPU core  == 1.73% of 8 CPU cores) of the CPU, which is less than the POD CPU limit of 12.5%. There are two possible reasons.**
+
+- APP A and Kmesh eBPF share the POD CPU limit, with Kmesh eBPF CPU being restricted by the POD CPU limit.
+- It is possible that due to the high performance of eBPF, APP A is not generating enough load to cause eBPF to exceed the CPU limit, further experiments in section 3.2 are needed.
```
Comment (hzxuzhonghu) [inline content/en/docs/performance/resource_consumption.md:56]: You can say this indicates ebpf cpu cost is fairly small compared to application itself.

I am also unsure if there are other applications in the same node sharing the bpf prog

**Review by hzxuzhonghu — APPROVED** [review]

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**